### ISSUE TITLE
25.0.3: systemctl reload docker.service doesn't wait for reload to finish

### ISSUE THREAD

**Issue by xinyi-joffre** [moby/moby#47353]
### Description

We have been using these two lines in our install scripts for past year without issues. It looks like as of yesterday, there is a race condition where we can no longer count on the configuration to be loaded when the systemctl command returns.

**Before, no issues running commands back to back:**
<install runsc>...
systemctl reload docker.service
docker run --rm --runtime=runsc hello-world

**In past few days, it is now consistently failing if we run the commands back to back:**
docker: Error response from daemon: unknown or invalid runtime name: runsc

If we add a sleep for a few seconds, then it can find the runsc runtime okay.

Looking at the systemctl logs, it seems like the systemd thinks reload is done, but the docker runtime is reporting reload success afterwards:
Feb 07 15:55:53 systemd[1]: Started Docker Application Container Engine.
Feb 07 15:55:58 systemd[1]: Reloading Docker Application Container Engine.
Feb 07 15:55:58 dockerd[7004]: time="2024-02-07T15:55:58.156917959Z" level=info msg="Got signal to relo>
Feb 07 15:55:58 systemd[1]: **Reloaded Docker Application Container Engine.**
Feb 07 15:55:58 dockerd[7004]: time="2024-02-07T15:55:58.214678240Z" level=info msg="**Reloaded configuration**>

Did something change with how `systemctl reload docker.service` behaves recently to consistently have this behavior? Or possibly some other reason the race condition has gotten worse, if it existed already?

### Reproduce

Ensure docker version is 25.0.3:

1. Run `systemctl reload docker.service`
2. Run `systemctl status docker.service`
3. See from logs that systemctl command completes before docker reports reload is complete

### docker version

```bash
Client: Docker Engine - Community
Version:           25.0.3
API version:       1.44
Go version:        go1.21.6
Git commit:        4debf41
Built:             Tue Feb  6 21:14:17 2024
OS/Arch:           linux/amd64
Context:           default
 
Server: Docker Engine - Community
Engine:
  Version:          25.0.3
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       f417435
  Built:            Tue Feb  6 21:14:17 2024
  OS/Arch:          linux/amd64
  Experimental:     false
containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
Version:    25.0.3
Context:    default
Debug Mode: false
Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.5
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
 
Server:
Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
Images: 0
Server Version: 25.0.3
Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: false
  userxattr: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Cgroup Version: 1
Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
Swarm: inactive
Runtimes: io.containerd.runc.v2 runc runsc
Default Runtime: runc
Init Binary: docker-init
containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
runc version: v1.1.12-0-g51d5e94
init version: de40ad0
Security Options:
  apparmor
  seccomp
   Profile: builtin
Kernel Version: 5.15.0-1054-azure
Operating System: Ubuntu 20.04.6 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.34GiB
Name: resources-estimation-001000120
ID: 7e461b39-7cdf-4292-99ce-e1495d595b7d
Docker Root Dir: /var/lib/docker
Debug Mode: false
Experimental: false
Insecure Registries:
  127.0.0.0/8
Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
Hm, I don't think there's ever been a guarantee that the configuration would be reloaded immediately. The reload signals the daemon, which in its turn starts to reload the config. https://github.com/moby/moby/blob/9e075f3808a5a5fd54f0217ac114d3a02b16af2f/contrib/init/systemd/docker.service#L14

There's been some changes wrt config handling in recent releases, so possibly timing is slightly different. Do you know which version of docker "works" and which version introduced the change in behavior?

**Comment by xinyi-joffre** [comment]
It looks like it started consistently failing during our VM re-images only in the past 48 hours, so most likely, it would be in a release that shipped very recently in past week or so. We've since re-imaged all of our VMs to hotfix the install scripts, so I unfortunately don't have the exact version of docker where this was working, but up until this week, it was very reliable.

For now, we put a 10 second sleep in between to be safe, but interesting that it started failing very consistently!

**Comment by cpuguy83** [comment]
Yes possibly related to refactoring, however this has always been impossible due to it being implemented using signals, which are asynchronous.

**Comment by xinyi-joffre** [comment]
I see, okay, we will add some waits/sleeps! Thanks for confirming that it's not a regression (and that it never waited on completion).

**Comment by thaJeztah** [comment]
Oh! Forgot to post; 

So it looks like `Type=notify-reload` was added in systemd 253, and not supported by older versions. Also see https://github.com/systemd/systemd/issues/6162

> It is strongly recommended to set `ExecReload=` to a command that not only triggers a configuration reload of the daemon, but also synchronously waits for it to complete.

As reloading through a signal is common, the new `Type=notify-reload` was added in https://github.com/systemd/systemd/pull/25916.

Also worth noting from that last one; https://github.com/systemd/systemd/pull/25916#issuecomment-1383775480

>> (...) Why the new type notify-reload is necessary, instead of silently extending `Type=notify`?
> We need to know if services implement reload or not, so that we can implement things such as "systemctl reload-or-restart" properly. `Type=notify` means no reload is implemented (or implemented via `ExecReload=`), while `Type=notify-reload` indicates that reload is implemented.

The `Type=notify-reload` option does not require a `ExecReload`, but instead can use the [`ReloadSignal`](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Type=), which was also added in systemd 253, but (IIUC) must be accompanied with [MONOTONIC_USEC](https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#MONOTONIC_USEC=…), which is not currently supported by the go-systemd library.


We currently still need to support distros that do not provide systemd 253, so cannot use this new feature, but perhaps "RELOADING=1 / "READY=1" might be the best we can do currently, and _perhaps_ would still give systemd a bit more information.

**Comment by thaJeztah** [comment]
Opened a PR for discussion;

- https://github.com/moby/moby/pull/47358
